### DISCUSSION TITLE
Switches should have their own options to Color their Thumb and Tint for state On/Off

### DISCUSSION THREAD

**Discussion by MAUIoxo** [dotnet/maui#21863] [Ideas]
### Description

Systems like Android often use a different Colors for the Slider and Thumb when they are toggled on or off like these:

![grafik](https://github.com/dotnet/maui/assets/140683727/1c49f6eb-8bd3-4501-afa1-7bcc8c763e55)

Also on iOS it could be interesting to create such effects easily. It could be possible to create something with `OnPlatform`-Properties and then use certain properties in the ViewModel or where the `Switch` is bound to and then change states correspondingly in the Code-Behind.

```xaml
<Switch IsToggled="{Binding IsSelected, Mode=TwoWay}" ThumbColor="{OnPlatform iOS={StaticResource Gray100}, Android={StaticResource DarkOrange3}}" OnColor="{OnPlatform iOS={StaticResource DarkOrange3}, Android={StaticResource DarkOrange3_Light}}" VerticalOptions="Center" HorizontalOptions="Center" Margin="{OnPlatform Android='0, -5, 0, -5'}" />
```

This, obviously is not so intuitive and also not so easily possible than having separate properties for:

- ThumbOffColor
- ThumbOnColor
- SliderOffColor (or TintOffColor)
- SliderOnColor (or TintOnColor)

So, I would propose to have certain properties for them and set them in `XAML` separately via these properties.

Another proposal would be that the right Color is somehow calculated internally which on **Android** is often a bit lighter for the the Slider/Tint than the acutal ThumbOnColor:

![grafik](https://github.com/dotnet/maui/assets/140683727/da37711d-d81c-492f-927e-d6a70e1b6167)

So, when a certain Color is set (e.g. orange) the right lighter version of that orange is calculated internally so that it fits the standards on Android.

Anyway, defining Colors for values below would be nice to have:

- ThumbOffColor
- ThumbOnColor
- SliderOffColor (or TintOffColor)
- SliderOnColor (or TintOnColor)

### Public API Changes

API Changes for the `Switch`

### Intended Use-Case

Use case is that developers can set the Colors for the following properties separately:

- ThumbOffColor
- ThumbOnColor
- SliderOffColor (or TintOffColor)
- SliderOnColor (or TintOnColor)

**Comment by jfversluis** [comment]
There already area styling options available for this, were you aware of those? Are those not sufficient? Why not?

See: https://learn.microsoft.com/dotnet/maui/user-interface/controls/switch

**Reply by bruno-gerardo** [reply]
When the device is in Dark Mode and you have a switch inside a layout with `BackgroundColor = "White"` switch's Off track color it's also set to White, and you only can see the Thumb.

**Reply by TypicalArthur** [reply]
Simply put, no.
Why would something like this not be editable? The track colour should be editable in both the on and off states, not just in the on state.

**Comment by MAUIoxo** [comment]
Sorry, I wasn't really aware of these styling options via `VisualStateGroups` for that. Still would prefer properties for these options as it would make it more obvious and clearer. But, can understand when you say it is better that you close this proposal.

**Comment by AlleSchonWeg** [comment]
Hi @jfversluis 
i miss the off track color.

**Reply by jfversluis** [reply]
I think `TrackColor` works with both on and off. Then toggle the color depending on the on/off state and you should be able to implement this functionality.

**Reply by MAUIoxo** [reply]
In general, after working with `XAML` for a while, you get into a mode so that you expect certain behaviors as Properties. I'm just saying that you would rather try to set something via "TrackOnColor" and "TrackOffColor" than setting big state groups:

```xaml
<Switch IsToggled="True">
    <VisualStateManager.VisualStateGroups>
        <VisualStateGroup x:Name="CommonStates">
            <VisualState x:Name="On">
                <VisualState.Setters>
                    <Setter Property="ThumbColor"
                            Value="MediumSpringGreen" />
                </VisualState.Setters>
            </VisualState>
            <VisualState x:Name="Off">
                <VisualState.Setters>
                    <Setter Property="ThumbColor"
                            Value="Red" />
                </VisualState.Setters>
            </VisualState>
        </VisualStateGroup>
    </VisualStateManager.VisualStateGroups>
</Switch>
```

This is just a general thing and preference for me. It is also more intuitive to browse the IntelliSense within a Control in XAML for certain properties.

So if you come to the point where you discuss internally how something new is going to be implemented 😊

**Reply by AlleSchonWeg** [reply]
Hi @jfversluis  and @MAUIoxo ,
the ThumbColor works. But the switch has only a OnColor property which colours the track. Using this VisualState:
```XAML
  <Switch IsToggled="True"
          Margin="10,0,15,0"
          VerticalOptions="Center">
    <VisualStateManager.VisualStateGroups>
      <VisualStateGroup x:Name="CommonStates">
        <VisualState x:Name="On">
          <VisualState.Setters>
            <Setter Property="ThumbColor"
                    Value="MediumSpringGreen" />
            <Setter Property="OnColor"
                    Value="MediumSpringGreen" />
          </VisualState.Setters>
        </VisualState>
        <VisualState x:Name="Off">
          <VisualState.Setters>
            <Setter Property="ThumbColor"
                    Value="Red" />
            <Setter Property="OnColor"
                    Value="Red" />
          </VisualState.Setters>
        </VisualState>
      </VisualStateGroup>
    </VisualStateManager.VisualStateGroups>
  </Switch>
```
results in a grey track if off:
![image](https://github.com/dotnet/maui/assets/6727366/f95d70fa-6f3e-4210-b734-86741e107a6b)

**Reply by MAUIoxo** [reply]
My preference is still, to have "OnThumbColor", "OffThumbColor", "OnTrackColor" and "OffTrackColor" - most obvious to me and regarding a clean API clearest

**Comment by TheXenocide** [comment]
Is there no way to control the color of the track when the switch is off? On some devices (e.g. a Pixel 6 Pro) it seems the default color has no visible contrast against a white background. The pixel emulators I've tried with seem to have a visible enough gray, but the physical device does not.

**Reply by chrond** [reply]
I have run into this issue on Android. When the background color is black, the track color is also black, so you can only see the thumb. On Windows, this is not a problem because the switch control has an outline around it.

**Reply by chrond** [reply]
Update: I found that setting the DefaultNightMode to ModeNightYes in the Android application code makes the track tint lighter so that you can see it on a dark background. I was previously just setting the UserAppTheme. It seems that if you can update the Android DefaultNightMode to match the UserAppTheme, this will solve the track visibility issue.

**Comment by stephenquan** [comment]
[REWRITE, 19th March 2025].

You can use BackgroundColor and Clip to setup the off color:

```xaml
 <Switch BackgroundColor="Red" OnColor="Green" ThumbColor="Orange">
     <Switch.Clip>
         <RoundRectangleGeometry CornerRadius="10" Rect="0,10,40,20" />
     </Switch.Clip>
</Switch>
```

![SwitchDemo.gif](https://github.com/user-attachments/assets/82cd5a10-53bb-4233-89ee-341932dc6667)

**Reply by TheXenocide** [reply]
Thanks for the sample. It's less than ideal to have to write our own just to address built-in behavior producing an invisible track, but it wouldn't be the first time.

**Reply by stephenquan** [reply]
@TheXenocide I rewrote my answer deleting the massive customization. The answer now shows a minimalistic way to achieve an off color.

**Reply by chrond** [reply]
Using the BackgroundColor is a very smart idea since we normally don't use that for a switch! It looks great on Windows. Unfortunately, this might not work for Android because the Thumb is larger than the track, so Clip will also cut off part of the thumb. (We'd also need different sizes and shapes of the track for each platform.)
![Switches](https://github.com/user-attachments/assets/f399f537-51e0-4e27-8688-382191fafbcd)

**Comment by nshtinkov** [comment]
I would like to see that, too. It is extremely annoying to have a switch with an invisible track and only a thumb. 
Note that (as others also mentioned) this happens only when the switch is "Off" in dark mode. Many of the examples only give an "On" switch and that does not show the problem.

**Reply by gchapeaux** [reply]
Just to give some extra info, there is actually no such thing as a "track color" in dark mode, but actually a little shading effect applied to the background, which ends up being hard to see on dark mode (try to shade some already dark color...)
Screenshot to illustrate, I added a red background to the first switch :

![image](https://github.com/user-attachments/assets/61e0b82c-47b4-4add-9470-b9ca3b17794d)

**Reply by chrond** [reply]
We either need a TrackColor property which would replace OnColor and can be handled by the on and off Visual States, or add an OffColor property in addition to the OnColor property.

**Reply by stephenquan** [reply]
@chrond I rewrote my answer to demonstrate how you can get OffColor. https://github.com/dotnet/maui/discussions/21863#discussioncomment-10271671

**Reply by chrond** [reply]
@stephenquan That's a slick hack. Thanks for the idea!

**Reply by chrond** [reply]
It's actually coming in .NET 10!
https://github.com/dotnet/core/blob/main/release-notes/10.0/preview/preview2/dotnetmaui.md#switchoffcolor

**Comment by Pastajello** [comment]
Developers hate this one simple trick!  Learn how:

https://github.com/user-attachments/assets/c44b9fdf-f09d-4518-9d20-c2b2581d89b1

Task.Delay(5) to the rescue! 
```
protected override async void OnPropertyChanged(string propertyName = null)
    {
        base.OnPropertyChanged(propertyName);
#if __ANDROID__
        if (propertyName == CustomSwitch.IsToggledProperty.PropertyName)
        {
            var view = Handler.PlatformView as AndroidX.AppCompat.Widget.SwitchCompat;
            if (IsToggled)
            {
                view.TrackDrawable?.SetColorFilter(OnColor.ToPlatform(), FilterMode.SrcAtop);
            }
            else  if(OffColor != default)
            {
                    await Task.Delay(5);
                    view.TrackDrawable?.SetColorFilter(OffColor.ToPlatform(), FilterMode.SrcAtop);
            }
        }
#endif
    }
```

**Comment by v0xw** [comment]
it's absolutely insane that this isn't an option. We need an OffColor or a TrackColor property which we can then change via visualstates.

**Comment by kdbhalala** [comment]
Anyone have workaround for it that works on Android and iOS?
